### PR TITLE
Replaced #ifndef header guards by #pragma once

### LINKED ISSUE #3774: Replace header guards of type `#ifndef` with a `#pragma once`
Replace `#ifndef` header guard with a `#pragma once`

_Originally posted by @kunaltyagi in https://github.com/_render_node/MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ0Nzg4MDI0OnYy/pull_request_review_threads/discussion_

### PR DESCRIPTION
**Pull request by 0sidharth**
Closes #3774
I have used a script which is given on this link -
https://github.com/marcusmueller/include-guard-convert
It covered 90% of the cases, however I had to manually edit the cases when #ifndef header guard is not the first line after comments.

### CODE REVIEW

File: apps/cloud_composer/include/pcl/apps/cloud_composer/tools/impl/organized_segmentation.hpp [hunk 1]
```diff
@@ -174,8 +173,4 @@ pcl::cloud_composer::OrganizedSegmentationTool::performTemplatedAction (const QL
 }
 
 
-#define PCL_INSTANTIATE_performTemplatedAction(T) template void pcl::cloud_composer::OrganizedSegmentationTool::performTemplatedAction<T> (QList <const CloudComposerItem*>);
-
-
-
-#endif //IMPL_TRANSFORM_CLOUDS_HPP_
+#define PCL_INSTANTIATE_performTemplatedAction(T) template void pcl::cloud_composer::OrganizedSegmentationTool::performTemplatedAction<T> (QList <const CloudComposerItem*>);
```
Comment (kunaltyagi) [inline apps/cloud_composer/include/pcl/apps/cloud_composer/tools/impl/organized_segmentation.hpp]: New line at the end of the file

File: common/include/pcl/common/impl/accumulators.hpp [hunk 2]
```diff
@@ -35,8 +35,7 @@
  *
  */
 
-#ifndef PCL_COMMON_IMPL_DETAIL_ACCUMULATORS_HPP
-#define PCL_COMMON_IMPL_DETAIL_ACCUMULATORS_HPP
+#pragma once
```
Comment (kunaltyagi) [inline common/include/pcl/common/impl/accumulators.hpp]: Please be consistent in either leaving an empty line or not between the top comment and pragma decl

Comment (0sidharth) [reply common/include/pcl/common/impl/accumulators.hpp]: Okay, I had simply run a script so the places that already had a space originally still have them. Will change it to leave a space always.

Comment (kunaltyagi) [reply common/include/pcl/common/impl/accumulators.hpp]: Got it. If you share the script, it'll be easier to review your PR

Comment (0sidharth) [reply common/include/pcl/common/impl/accumulators.hpp]: I have updated the description of this PR with the link.

**Comment by 0sidharth** [comment]
I am running Ubuntu 16.04 and it successfully compiled for me, why is the build here failing? I am getting a warning though, about using pragma once in a main file (ml/src/ svm_wrapper.cpp and svm .cpp which originally had #ifndefs".

File: apps/cloud_composer/include/pcl/apps/cloud_composer/tools/impl/organized_segmentation.hpp (hunk 1 above)
Comment (0sidharth) [reply apps/cloud_composer/include/pcl/apps/cloud_composer/tools/impl/organized_segmentation.hpp]: Are you saying the file need an empty line at the end? The line at the end was originally there

**Comment by kunaltyagi** [comment]
> using pragma once in a main file

Don't do that. Those defines should be removed. 16.04 treats warnings as errors so that explains it

**Comment by 0sidharth** [comment]
So must I make the code uniform in terms of new lines between pragma decl and comments?
apps/modeler/include/pcl/apps/modeler/impl/parameter.hpp
2d/include/pcl/2d/impl/convolution.hpp
these are examples of 2 files that you can see already had differing formats.

**Comment by kunaltyagi** [comment]
You can modify the tool you used. If you use the following regex (to find preceding empty lines) instead of what's present on L16, the tool should remove the empty line between the comment and the pragma (assuming you convert from master again)
```
		'ifndef' : '[^\s*$]*^\s*#(?:ifndef|IFNDEF)\s+([A-Za-z_0-9]{4,})\s*$',
```

**Comment by 0sidharth** [comment]
Are you talking about this?
```
regexes = {
		'ifndef' : '[^\s*$]*^\s*#(?:ifndef|IFNDEF)\s+([A-Za-z_0-9]{4,})\s*$',
		'define' : '^\s*#(?:define|DEFINE)\s+([A-Za-z_0-9]{4,})\s*$',
		'endif' : '^\s*#(?:endif|ENDIF)\s*(/\*.*\*/|//.+)?\s*$',
		'blank' : '^\s*(/\*.*\*/|//.+)?\s*$',
		'pragma' : '^\s*#(?:pragma|PRAGMA)\s+(?:once|ONCE)'
}
```
Does not work

**Comment by 0sidharth** [comment]
Could you please clarify what I need to do?

**Comment by kunaltyagi** [comment]
1. Run the tool normally to get the weird mix (current status)
2. Run sed on all files: `find . -t f -exec sed -i 's/^\s*#pragma once/#pragma once/' \;` or similar command to remove all empty lines before a `#pragma once`
3. Ensure the CI is green, not red
4. PR-o-fit

**Comment by 0sidharth** [comment]
Well so I tried around with sed a bit and just couldn't get it work. So I ended up writing a python script to remove the extra line, the script is as below
```
import os

path = '/home/sidharth/Desktop/pcl/'
files = []

for r, d, f in os.walk(path):
    for file in f:
        if '.h' in file:
            files.append(os.path.join(r, file))

flag = False
prevLine = ''
for f in files:
    with open(f, "r") as f_in:
        lines = f_in.readlines()
    with open(f, "w") as f_out:
        for line in lines:
            if flag is True:
                if line == '#pragma once\n':
                    print 'here'
                    f_out.write(line)
                    flag = False
                else:
                    f_out.write(prevLine)
                    f_out.write(line)
                    flag = False
            elif line in ['\n', '\r\n']:
                flag = True
                prevLine = line
            else:
                flag = False
                f_out.write(line)
```
This worked (I think), compiling and commiting. Can you check this script in the meanwhile (as in the previous case, I think this would make it easier to review)?

File: 2d/include/pcl/2d/impl/kernel.hpp [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (kunaltyagi) [inline 2d/include/pcl/2d/impl/kernel.hpp]: Empty line at the beginning of the file

File: apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/utils/vtk_model_sampling.h [hunk 4]
```diff
@@ -1,10 +1,10 @@
+
```
Comment (kunaltyagi) [inline apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/utils/vtk_model_sampling.h]: Same here

File: apps/cloud_composer/include/pcl/apps/cloud_composer/impl/cloud_item.hpp [hunk 5]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (kunaltyagi) [inline apps/cloud_composer/include/pcl/apps/cloud_composer/impl/cloud_item.hpp]: And here

**Comment by kunaltyagi** [comment]
New code snippet please 😄

**Comment by 0sidharth** [comment]
Oh, well actually I took the easy way out and just scanned and removed the extra line at the start of the files following my last commit using
```
import os

path = '/home/sidharth/Desktop/pcl/'
files = []

for r, d, f in os.walk(path):
    for file in f:
        if '.h' in file:
            files.append(os.path.join(r, file))

flag = False
prevLine = ''
count = 0
for f in files:
    count = 0
    with open(f, "r") as f_in:
        lines = f_in.readlines()
    with open(f, "w") as f_out:
        for line in lines:
            count = count + 1
            if count == 1:
                if line in ['\n', '\r\n']:
                    flag = True
                else:
                    f_out.write(line)
            else:
                f_out.write(line)
```

BTW, I am getting some compile errors due to redifinition of struct in the CI, no errors on my device tho. After researching a bit, I suspect it is due to missing header guards in the CPP files. So I have temporarily added back the `ifndef` header guards to the 3 cpp files

**Comment by kunaltyagi** [comment]
2 issues:
```
/__w/1/s/surface/include/pcl/surface/3rdparty/opennurbs/deflate.h:92:16: error: redefinition of 'struct internal_state'
 typedef struct internal_state {
                ^
In file included from /__w/1/s/surface/include/pcl/surface/3rdparty/opennurbs/zutil.h:16:0,
                 from /__w/1/s/surface/include/pcl/surface/3rdparty/opennurbs/deflate.h:14,
                 from /__w/1/s/surface/src/3rdparty/opennurbs/deflate.c:52:
/__w/1/s/surface/include/pcl/surface/3rdparty/opennurbs/zlib.h:78:8: note: originally defined here
 struct internal_state;
        ^
```
and
```
In file included from /__w/1/s/surface/src/mls.cpp:41:0:
/__w/1/s/surface/include/pcl/surface/impl/mls.hpp: In member function 'void pcl::MLSResult::computeMLSSurface(const pcl::PointCloud<PointT>&, int, const std::vector<int>&, double, int, std::function<double(double)>) [with PointT = pcl::PointXYZ]':
/__w/1/s/surface/include/pcl/surface/impl/mls.hpp:723:19: warning: 'xyz_centroid' may be used uninitialized in this function [-Wmaybe-uninitialized]
   Eigen::Vector4d xyz_centroid;
                   ^
```

**Comment by 0sidharth** [comment]
Okay, will try to fix those. Is there any reason why the build was still successful on Mac? Does it have anything to do with compiler versions?

**Comment by kunaltyagi** [comment]
> any reason why the build was still successful on Mac?

Surface module is switched off on Mac. I don't know the reason though. @taketwo might

**Comment by taketwo** [comment]
Sure, I do. It's off... because you switched it off, Kunal :rofl: https://github.com/PointCloudLibrary/pcl/commit/6f1932853fc5e9f0820471ed158a23891874dfbe

**Comment by kunaltyagi** [comment]
Oh the irony...

**Comment by kunaltyagi** [comment]
@0sidharth The CI are red

**Comment by kunaltyagi** [comment]
Pinging and marking as stale

**Comment by mvieth** [comment]
I'm closing this pull request because there are too many merge conflicts -- it's probably easier to start new instead of trying to fix all

**Pull request closed without merging**